packages/docs/docs/docs/model/01-definition.md

```diff
@@ -1,284 +1,382 @@
+// logged in, or a guest
+const user = types.union({
+  loggedIn: types.object({ name: types.string() }),
+  guest: types.object({}),
+})
+```
+
+#### A thorough example
+
+With these building blocks, we have an powerful toolbox to expressively describe complex domains.
+
+Let's work through a more complex example and see how this would work out.
+In this example we're modelling (a simplified version of) user login:
+
+- we receive a password and username as input
+- we send back a response that can either be successful of contain an error message:
+  - if the user can be logged in the response is successful and contains the user information
+  - if the user cannot be logged in the response will contain an error message
+
+The type definitions needed for this example would be the following:
 
 ```ts showLineNumbers
- // two equivalent defintions
-const ArrayOfStrings = m.array(m.string())
-const ArrayOfStrings = m.string().array()
+// The type of users, arguably it could be more complex but it's ok as an example
+type User = types.Infer<typeof user>
+const user = types.object({
+  id: types.integer(),
+  username: types.string(),
+})
 
-const ArrayOfMaxFiveStrings = m.string().array({ maxItems: 5 })
+// The input we receive when a user wants to login
+type AuthenticationData = types.Infer<typeof authenticationData>
+const authenticationData = types.object({
+  username: types.string(),
+  password: types.string(),
+})
 
-const User = m.object({
-  id: m.integer(),
-  name: m.string(),
-  surname: m.string(),
-  // highlight-start
-  emails: m.string().array(),
-  // highlight-end
+// The type of the response: it can either be a `success` or `failure`
+type LoginResponse = types.Infer<typeof loginResponse>
+const loginResponse = types.union({
+  success: user,
+  failure: types.object({
+    reason: types.string(),
+  })
 })
 ```
 
-Combining the `array` decorator with others like `optional`, `nullable`, and `default` assumes different meanings depending on the order in which they are applied. Note, for example, the following two cases: 
+Given these type definitions we can get an idea of how the login process could work out:
 
 ```ts showLineNumbers
-import { m, decode } from '@mondrian-framework/model'
+async function loginUser(auth: AuthenticationData) {
+  const user = await fetchUser(auth.username, auth.password)
+  const response = user
+    ? { success: user }
+    : { failure: { reason: "wrong username or password" } }
 
-const NullableArrayOfStrings = m.string().array().nullable()
-const ArrayOfNullableStrings = m.string().nullable().array()
+  await sendResponse(response)
+}
+```
 
-decode(NullableArrayOfStrings, null) // => null
-decode(NullableArrayOfStrings, [null]) // => error
+Types definitions can help us define clear and expressive models that are faithful to the modelled domain.
 
-decode(ArrayOfNullableStrings, null) // => error
-decode(ArrayOfNullableStrings, [null]) // => [null]
+## Additional Types
```

Comment:
I think you forgot datetime, timestamp, void and null, that were originally listed in the primitives and now, I suppose, are additional types. 

Besides, I know additional types are in theory less interesting than the previous, but from a user perspective types like datetime and timestamp are missing until they reach this section that is far far below the top. We should consider moving them up, what do you think?

Reply:
Yes it makes perfect sense, I'll move the additional types to the top right below the primitives.

Now that I think about it there's also a datetime in `model` but I forgot about it, I'll add it back as well